firedrake/meshadapt.py

```diff
@@ -0,0 +1,113 @@
+from __future__ import absolute_import, print_function, division
+from six import with_metaclass
+import abc
+
+import numpy as np
+
+from firedrake.mesh import Mesh
+import firedrake.dmplex as dmplex
+import firedrake.utils as utils
+import firedrake.functionspace as functionspace
+import firedrake.function as function
+import firedrake.mesh as fmesh
+
+
+__all__ = ['adapt']
+
+
+class BaseAdaptation(with_metaclass(abc.ABCMeta)):
+    """
+    Abstract top-level class for mesh adaptation
+    """
+
+    def __init__(self, mesh):
+        """
+        """
+        self.mesh = mesh
+
+    @abc.abstractproperty
+    def newmesh(self):
+        pass
+
+    @abc.abstractmethod
+    def transfer_solution(self, f, **kwargs):
+        pass
+
+
+class AAdaptation(BaseAdaptation):
+    """
+    Object that performs anisotropic mesh adaptation
+    """
+
+    def __init__(self, mesh, metric):
+        """
+        """
+        if isinstance(mesh.topology, fmesh.ExtrudedMeshTopology):
+            raise NotImplementedError("Cannot adapt extruded meshes")
+        if mesh.coordinates.ufl_element().family() != 'Lagrange' \
+           or mesh.coordinates.ufl_element().degree() != 1:
+            raise NotImplementedError("Mesh coordinates must be P1.")
+        if metric.ufl_element().family() != 'Lagrange' \
+           or metric.ufl_element().degree() != 1:
+            raise ValueError("Metric should be a P1 field.")
+        super(AAdaptation, self).__init__(mesh)
+        self.metric = metric
+
+    @utils.cached_property
+    def newmesh(self):
+        """
+        Generates the adapted mesh wrt the metric
+        """
+        plex = self.mesh.topology._plex
+        dim = self.mesh._topological_dimension
+        # PETSc adapt routine expects that the right coordinate section is set
+        # hence the following bloc of code
+        entity_dofs = np.zeros(dim+1, dtype=np.int32)
+        entity_dofs[0] = self.mesh.geometric_dimension()
+        coordSection = plex.createSection([1], entity_dofs, perm=self.mesh.topology._plex_renumbering)
+        dmCoords = plex.getCoordinateDM()
+        dmCoords.setDefaultSection(coordSection)
+
+        with self.mesh.coordinates.dat.vec_ro as coords:
+            plex.setCoordinatesLocal(coords)
```

Comment:
The change you suggest does not work (in serial, the mesh transfered to pragmatic is wrong) whereas current version works. I am not entirely sure I understand why your version would be better ?